### PR TITLE
Add exception to problem details context

### LINKED ISSUE #47651: Add exception property to the problem details context
## Background and Motivation

If an application is configured to generate problem details it would be good, if an exception occurs in the application, that the original exception can be retreived directly on the problem details context. The exception information can be used to customize the problem details when using the CustomizeProblemDetails functionality. The original exception can already be retreived using the IExceptionHandlerFeature, but having the exception directly on the context will be easier and more straight forward to use.

The exception property should be set when the problem details context is written from the built-in exception handler middlewares.

## Proposed API

```diff
namespace Microsoft.AspNetCore.Http;

public class ProblemDetailsContext
{
+   public Exception? Exception { get; init; }
}

```

## Usage Examples

```csharp
builder.Services.AddProblemDetails(configure =>
    configure.CustomizeProblemDetails = (context) =>
    {
        if (context.Exception is not null)
        {
            // Customize problem details with information from exception.
        }
    }
);
```

## Alternative Designs

N/A

## Risks

None that I am aware of.

### PR DESCRIPTION
**Pull request by bjornen77**
# Add exception to problem details context

Adds an exception property to the problem details context. The exception is set in the developer exception page and exception handler middlewares.

Fixes #47651

### CODE REVIEW

File: src/Middleware/Diagnostics/src/DeveloperExceptionPage/DeveloperExceptionPageMiddlewareImpl.cs [hunk 1]
```diff
@@ -205,8 +205,8 @@ private async Task DisplayExceptionContent(ErrorContext errorContext)
             SetExceptionHandlerFeatures(errorContext);
         }
 
-        if (_problemDetailsService == null ||
-            !await _problemDetailsService.TryWriteAsync(new() { HttpContext = httpContext, ProblemDetails = CreateProblemDetails(errorContext, httpContext) }))
+        if (_problemDetailsService == null || !await _problemDetailsService.TryWriteAsync(new()
+            { HttpContext = httpContext, ProblemDetails = CreateProblemDetails(errorContext, httpContext), Exception = errorContext.Exception }))
```
Comment (Tratcher) [inline src/Middleware/Diagnostics/src/DeveloperExceptionPage/DeveloperExceptionPageMiddlewareImpl.cs:209]:
```suggestion
            {
                HttpContext = httpContext,
                ProblemDetails = CreateProblemDetails(errorContext, httpContext), 
                Exception = errorContext.Exception 
            }))
```

**Review by captainsafia — APPROVED** [review]

**Comment by JamesNK** [comment]
The PR's linked issue has details.

**Comment by niemyjski** [comment]
Just seeing this now, but I feel like it's a bit of a mistake that CustomizeProblemdetails isn't passed the Exception instance by the default writer. Because I want to extend the problem details based on error type.
```
if (ctx.Exception is MiniValidatorException validationException)
{
    var invalidParams = validationException.Errors.SelectMany(e => e.Value.Select(message => new ValidationProblemDetailsParam(e.Key.ToLowerUnderscoredWords(), message))).ToList();
    ctx.ProblemDetails.Extensions.Add("invalid-params", invalidParams);
}```

**Comment by bjornen77** [comment]
> Just seeing this now, but I feel like it's a bit of a mistake that CustomizeProblemDetails isn't passed the Exception instance by the default writer. Because I want to extend the problem details based on error type.
> 
> ```cs
> if (ctx.Exception is MiniValidatorException validationException)
> {
>     var invalidParams = validationException.Errors.SelectMany(e => e.Value.Select(message => new ValidationProblemDetailsParam(e.Key.ToLowerUnderscoredWords(), message))).ToList();
>     ctx.ProblemDetails.Extensions.Add("invalid-params", invalidParams);
> }
> ```

Not sure i understand exactly what you mean. But if you use the default exception handler "app.UseExceptionHandler();" the exception should be available on the ctx.Exception in the code above.

**Comment by niemyjski** [comment]
It's not, hence I added this comment.

**Comment by bjornen77** [comment]
Strange, it should work. I do not see anything obvious in your code that could cause this.

One thing to keep in mind is that the exception is only available if the problem is caused by an unhandled exception. For example, it is not available if a Problem() is returned manually from an API controller or if you have an IExceptionHandler that handles the exception(return true). Also make sure that the response have not started already.

This setting could also effect if an exception is thrown on bad requests or not:

builder.Services.Configure<RouteHandlerOptions>(options =>
{
    options.ThrowOnBadRequest = true;
});

**Pull request merged**